text/0059-nodes-capabilities-discovery.md

```diff
@@ -0,0 +1,142 @@
+# RFC-0059: Add a discovery mechanism for nodes based on their capabilities
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 2023-12-18                                                                                  |
+| **Description** | Nodes having certain capabilities register themselves in the DHT to be discoverable         |
+| **Authors**     | Pierre Krieger                                                                              |
+
+## Summary
+
+This RFC proposes to make the mechanism of [RFC #8](https://github.com/polkadot-fellows/RFCs/blob/main/text/0008-parachain-bootnodes-dht.md) more generic by introducing the concept of "capabilities".
+
+Implementations can implement certain "capabilities", such as serving old block headers or being a parachain bootnode.
+
+The discovery mechanism of RFC #8 is extended to be able to discover nodes of specific capabilities.
+
+## Motivation
+
+The Polkadot peer-to-peer network is made of nodes. Not all these nodes are equal. Some nodes store only the headers of recently blocks, some nodes store all the block headers and bodies since the genesis, some nodes store the storage of all blocks since the genesis, and so on.
+
+It is currently not possible to know ahead of time (without connecting to it and asking) which nodes have which data available, and it is not easily possible to build a list of nodes that have a specific piece of data available.
+
+If you want to download for example the header of block 500, you have to connect to a randomly-chosen node, ask it for block 500, and if it says that it doesn't have the block, disconnect and try another randomly-chosen node.
+In certain situations such as downloading the storage of old blocks, nodes that have the information are relatively rare, and finding through trial and error a node that has the data can take a long time.
+
+This RFC attempts to solve this problem by giving the possibility to build a list of nodes that are capable of serving specific data.
+
+## Stakeholders
+
+Low-level client developers.
+People interested in accessing the archive of the chain.
+
+## Explanation
+
+*Reading RFC #8 first might help with comprehension, as this RFC is very similar.*
+
+Please keep in mind while reading that everything below applies for both relay chains and parachains, except mentioned otherwise.
+
+### Capabilities
+
+This RFC defines a list of so-called **capabilities**:
```

Comment:
worth clarifying that this is non-exhaustive and that future capabilities may be added.